### PR TITLE
[Data] Fix `PipelineIterator.to_tf`

### LINKED ISSUE #32528: [Datasets] `PipelinedDatasetIterator.to_tf` doesn't work
### What happened + What you expected to happen

I tried to train a TensorFlow model with per-epoch preprocessing, but I got an error:

```
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "/Users/balaji/Documents/GitHub/ray/.venv/lib/python3.10/site-packages/tensorflow/python/data/ops/iterator_ops.py", line 787, in __next__
    return self._next_internal()
  File "/Users/balaji/Documents/GitHub/ray/.venv/lib/python3.10/site-packages/tensorflow/python/data/ops/iterator_ops.py", line 770, in _next_internal
    ret = gen_dataset_ops.iterator_get_next(
  File "/Users/balaji/Documents/GitHub/ray/.venv/lib/python3.10/site-packages/tensorflow/python/ops/gen_dataset_ops.py", line 3017, in iterator_get_next
    _ops.raise_from_not_ok_status(e, name)
  File "/Users/balaji/Documents/GitHub/ray/.venv/lib/python3.10/site-packages/tensorflow/python/framework/ops.py", line 7215, in raise_from_not_ok_status
    raise core._status_to_exception(e) from None  # pylint: disable=protected-access
tensorflow.python.framework.errors_impl.UnknownError: {{function_node __wrapped__IteratorGetNext_output_types_2_device_/job:localhost/replica:0/task:0/device:CPU:0}} RuntimeError: Pipeline cannot be read multiple times.
Traceback (most recent call last):

  File "/Users/balaji/Documents/GitHub/ray/.venv/lib/python3.10/site-packages/tensorflow/python/ops/script_ops.py", line 271, in __call__
    ret = func(*args)

  File "/Users/balaji/Documents/GitHub/ray/.venv/lib/python3.10/site-packages/tensorflow/python/autograph/impl/api.py", line 642, in wrapper
    return func(*args, **kwargs)

  File "/Users/balaji/Documents/GitHub/ray/.venv/lib/python3.10/site-packages/tensorflow/python/data/ops/dataset_ops.py", line 1039, in generator_py_func
    values = next(generator_state.get_iterator(iterator_id))

  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/dataset.py", line 3368, in generator
    for batch in self.iter_batches(

  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/dataset_pipeline.py", line 219, in iter_batches
    raise RuntimeError("Pipeline cannot be read multiple times.")

RuntimeError: Pipeline cannot be read multiple times.


         [[{{node PyFunc}}]] [Op:IteratorGetNext]
```

### Versions / Dependencies

Ray: 66c05338b07f1ef149928d4742b5f70c6c49b138
Python: 3.9

### Reproduction script

```
import ray
from ray.data._internal.pipelined_dataset_iterator import PipelinedDatasetIterator


dataset = ray.data.from_items([{"x": 0, "y": 0}])
pipeline = dataset.repeat()
iterator = PipelinedDatasetIterator(pipeline)

tf_dataset = iterator.to_tf("x", "y")
for epoch in range(2):
    for (inputs, targets) in tf_dataset:
        pass
```

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by amogkam**
Signed-off-by: amogkam <amogkamsetty@yahoo.com>

Closes https://github.com/ray-project/ray/issues/32528. 

We move `to_tf` logic to `DatasetIterator`, leveraging the respective `iter_batches` implementations for `BulkDatasetIterator` and `PipelineDatasetIterator`.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/_internal/pipelined_dataset_iterator.py [hunk 1]
```diff
@@ -26,7 +26,9 @@ def __repr__(self) -> str:
     def _get_next_dataset(self) -> "Dataset":
         if self._epoch_iterator is None:
             self._epoch_iterator = self._base_dataset_pipeline.iter_epochs()
-        ds = next(self._epoch_iterator)
+
+        # The epoch iterator will always have 1 dataset.
```
Comment (bveeramani) [inline python/ray/data/_internal/pipelined_dataset_iterator.py:30]: Is this necessarily true? If we're streaming, then we call `dataset.window(bytes_per_window=stream_window_size).repeat()`. In this case, wouldn't there be multiple datasets in an epoch?

```
>>> dataset = ray.data.range(100000)
>>> pipeline = dataset.window(blocks_per_window=1).repeat()
>>> len(list(epoch.iter_datasets()))                                                             
20
```

File: python/ray/data/tests/test_dataset_iterator.py [hunk 2]
```diff
@@ -44,6 +44,29 @@ def test_tf_conversion(ray_start_regular_shared):
...
+    it = ds.iterator()
+    tf_dataset = it.to_tf("value", "value")
+    for i, row in enumerate(tf_dataset):
+        assert all(row[0] == i)
+        assert all(row[1] == i)
+        assert isinstance(row[0], tf.Tensor)
+        assert isinstance(row[1], tf.Tensor)
+
+    # Repeated twice.
+    tf_dataset = it.to_tf("value", "value")
```
Comment (bveeramani) [inline python/ray/data/tests/test_dataset_iterator.py:58]: Remove? If you're using TensorFlow, you'd only call `to_tf` once and re-use the same TensorFlow dataset.
```suggestion
```

File: python/ray/data/_internal/pipelined_dataset_iterator.py (hunk 1 above)
Comment (amogkam) [reply python/ray/data/_internal/pipelined_dataset_iterator.py:30]: where are you calling `iter_epochs` here?

File: python/ray/data/tests/test_dataset_iterator.py (hunk 2 above)
Comment (amogkam) [reply python/ray/data/tests/test_dataset_iterator.py:58]: This is to test that the `repeat` is being respected.

File: python/ray/data/_internal/pipelined_dataset_iterator.py (hunk 1 above)
Comment (amogkam) [reply python/ray/data/_internal/pipelined_dataset_iterator.py:29]: updated to not use `iter_dataset`

**Comment by amogkam** [comment]
Depends on #32526

File: python/ray/data/_internal/pipelined_dataset_iterator.py [hunk 3]
```diff
@@ -19,13 +17,15 @@ def __init__(
     ):
         self._base_dataset_pipeline = base_dataset_pipeline
         self._epoch_iterator = None
+        self._to_tf_called = False
```
Comment (clarkzinzow) [inline python/ray/data/_internal/pipelined_dataset_iterator.py:20]: I might have missed it, but why is this needed?

**Review by clarkzinzow — APPROVED** [review]

File: python/ray/data/_internal/pipelined_dataset_iterator.py (hunk 3 above)
Comment (amogkam) [reply python/ray/data/_internal/pipelined_dataset_iterator.py:20]: ah sorry, good catch. I will remove this, it is not needed.

**Pull request merged**